libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -314,6 +355,154 @@ bool IRGeneratorForStatements::visit(FunctionCall const& _functionCall)
+			defineExpression(_memberAccess) <<
+				"balance(" <<
+				expressionAsType(_memberAccess.expression(), *TypeProvider::address()) <<
+				")\n";
+		else if ((set<string>{"send", "transfer"}).count(member))
+		{
+			solAssert(dynamic_cast<AddressType const&>(*_memberAccess.expression().annotation().type).stateMutability() == StateMutability::Payable, "");
+			defineExpression(_memberAccess) <<
+				expressionAsType(_memberAccess.expression(), *TypeProvider::payableAddress()) <<
+				"\n";
+		}
+		else if ((set<string>{"call", "callcode", "delegatecall", "staticcall"}).count(member))
+			defineExpression(_memberAccess) <<
+				expressionAsType(_memberAccess.expression(), *TypeProvider::address()) <<
+				"\n";
+		else
+			solAssert(false, "Invalid member access to address");
+		break;
+	}
+	case Type::Category::Function:
+		if (member == "selector")
+		{
+			solUnimplementedAssert(false, "");
+		}
+		else
+			solAssert(
+				!!_memberAccess.expression().annotation().type->memberType(member),
+				"Invalid member access to function."
+			);
+		break;
+	case Type::Category::Magic:
+		// we can ignore the kind of magic and only look at the name of the member
+		if (member == "coinbase")
+			defineExpression(_memberAccess) << "coinbase()\n";
+		else if (member == "timestamp")
+			defineExpression(_memberAccess) << "timestamp()\n";
+		else if (member == "difficulty")
+			defineExpression(_memberAccess) << "difficulty()\n";
+		else if (member == "number")
+			defineExpression(_memberAccess) << "number()\n";
+		else if (member == "gaslimit")
+			defineExpression(_memberAccess) << "gaslimit()\n";
+		else if (member == "sender")
+			defineExpression(_memberAccess) << "caller()\n";
+		else if (member == "value")
+			defineExpression(_memberAccess) << "callvalue()\n";
+		else if (member == "origin")
+			defineExpression(_memberAccess) << "origin()\n";
+		else if (member == "gasprice")
+			defineExpression(_memberAccess) << "gasprice()\n";
+		else if (member == "data")
+			solUnimplementedAssert(false, "");
+		else if (member == "sig")
+			defineExpression(_memberAccess) <<
+				"and(calldataload(0), " <<
+				formatNumber(u256(0xffffffff) << (256 - 32)) <<
+				")\n";
+		else if (member == "gas")
+			solAssert(false, "Gas has been removed.");
+		else if (member == "blockhash")
+			solAssert(false, "Blockhash has been removed.");
+		else if (member == "creationCode" || member == "runtimeCode")
+		{
+			solUnimplementedAssert(false, "");
+		}
+		else if (member == "name")
+		{
+			solUnimplementedAssert(false, "");
+		}
+		else if ((set<string>{"encode", "encodePacked", "encodeWithSelector", "encodeWithSignature", "decode"}).count(member))
+		{
+			// no-op
+		}
+		else
+			solAssert(false, "Unknown magic member.");
+		break;
+	case Type::Category::Struct:
+	{
+		solUnimplementedAssert(false, "");
+	}
+	case Type::Category::Enum:
+	{
+		EnumType const& type = dynamic_cast<EnumType const&>(*_memberAccess.expression().annotation().type);
```

Comment:
Is this cast always true?

Reply:
If not, it will throw...

Reply:
This code was adapted from the classic code generator, so I would say it is fine.